# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [1]:
import pandas as pd

In [2]:
sep_lines = "\n"+"-"*75+"\n"

In [3]:
# Your code here
url =  "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
car_insurance_df = pd.read_csv(url)
rows, cols = car_insurance_df.shape
print(f"Shape: {rows} rows x {cols} columns", end=sep_lines)
print(car_insurance_df.dtypes, end=sep_lines)
print(car_insurance_df.nunique(),end=sep_lines)
print(car_insurance_df['Policy Type'].unique(),end=sep_lines)
print(car_insurance_df['Vehicle Class'].unique(),end=sep_lines)
print(car_insurance_df['Education'].unique(),end=sep_lines)
print(car_insurance_df['Number of Open Complaints'].unique(),end=sep_lines)

Shape: 4008 rows x 11 columns
---------------------------------------------------------------------------
Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object
---------------------------------------------------------------------------
Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64
----------

### **data types:**
**from reading the data types**:
It seems any categorical or string id columns are object type while the few numerical are float64.
2 Columns seem to have out of place dtypes:
- Number of Open Complaints: should be INT instead of object
- Customer Lifetime Value, which should be Float64 instead of object

### Data Categorization:
**From looking at unique value counts and data types of columns:**
- Numerical:
  - Continuous: Customer Lifetime value, Income, Monthly payment,Total Claim
  - Discrete: Number of open complaints

- Categorical:
  - Nominal: State, Gender, Vehicle Class(different type of cars (is a SUV higher ranking than a 2-4 door car? can't be answered so...))
  - Ordinal: Education, Policy type(assuming special > corporate > personal, otherwise it would be nominal.)





In [4]:
car_insurance_df.describe(include='number')

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

In [5]:
car_insurance_df.describe(include='object')

Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       Number of Open Complaints    Policy Type  Vehicle Class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576

In [6]:
import re
gender = set(["F","M"])
gender_s = car_insurance_df['GENDER'].value_counts()
for key in car_insurance_df['GENDER'].unique():
    if key in gender: continue
        
    if re.match('[Ff]', str(key)):
        gender_s["F"] += gender_s[key]
        gender_s.drop(key,inplace=True)

    if re.match('[mM]', str(key)):
        gender_s["M"] += gender_s[key]
        gender_s.drop(key,inplace=True)
    
gender_s['Sum'] = gender_s.sum()
gender_s['ratio F-M'] = f"{gender_s['F']/gender_s['Sum']*100:.2f}% - {gender_s['M']/gender_s['Sum']*100:.2f}%"
gender_s

GENDER
F                        502
M                        452
Sum                      954
ratio F-M    52.62% - 47.38%
Name: count, dtype: object

In [7]:
for key in car_insurance_df.describe(include='object').columns:
    if key != 'Customer':
        print(f"{key} unique: {car_insurance_df[key].unique()}")
        print(f"{key} unique: {car_insurance_df[key].value_counts()}",end=sep_lines)

ST unique: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
ST unique: ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64
---------------------------------------------------------------------------
GENDER unique: [nan 'F' 'M' 'Femal' 'Male' 'female']
GENDER unique: GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64
---------------------------------------------------------------------------
Education unique: ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
Education unique: Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64
---------------------------------------------------------------------------
Custome

### Data Descripiton + summary statistics:
  **Categorical:**
  - Policy type :  special auto, corporate auto, personal auto
    - according to the data, most clients have the personal auto plan, followed by corporate then special plans; seems to indicate most of their clients are not in a corporate offiece or relatively wealthy.

  - Education : High School/below, College, Bachelor, Master, Doctor (note: duplicate in Bachelors)
    - According to the data, it seems majority of clients went to college, although many didn't finish, some only have high school or below education, and only a few clients have advanced degrees. This could indicate that the insurance company mainly serves those with only a high school or undergraduate college education, working/middle class income workers.

  - Vehicle Class: Four-Door Car, Two-Door Car, SUV, Luxury SUV ,Sports Car, Luxury Car
    - Most policy holders have 4 door cars, with two door cars and SUVs taking the next portion, and luxury classes are last. This indicates that most clients don't own luxurious car vehicles and only some have access to SUVs for carrying  4+ passangers and plenty of cargo space.
  - Gender: Female, Male
    - Female 502 (52.62%) ; Male 452 (47.38%): indicates a slight lean towards Women clients. Could maybe explain some income discrepancies; women tend to make less income on avg and some stay home to take care of kids full-time as mothers, which could explain some zero income policy holders.
  - States: Washington Arizona Nevada California Oregon
    - 1st California 331 ; 2nd Oregon 320 ; 3rd Arizona 211; 4th Washington 111 5th Nevada 98 ; Seems to indicate car insurande mainly serves **South/North Western USA,** particularly in Cali and Oregon. Could indicate current market reach in insurance company.


**Insights**:
This car insurance company seems to provide policies for working/middle class clients, corporations/corporate clients, and wealthier clients. It seems they contain a slight majority of women, most are working class, most have 4 door/ 2 door cars, most have personal auto plans, clientale is based in Western USA, and that most have at least some college education or high school education.




**Numerical:**
 - Income: \$0 - \$99960.0: indicates that they accept a broad level of income level people, even those with no recorded income; however this could be explained by couple of things, like teens who get their car insurance paid for by their parents, or corporate offices not disclosing the income of their employees. However, it does seem that they don't not have any clients with 6 or more figure salaries.
    - mean: 39295.7
    - median: 36234.000000
    - std: 30469.427060
    - *Stat Conclusion*: The mean income is ~40k, indicating the average client is in the lower income class, the median being slightly lower indicates a slight skew of higher paying clientele, and the std indicates a seemingly very wide spread of incomes at around ~30k. This indicates that they have a majority of low income clients with a few relatively more wealthy clients.

 - Monthly payment: \$61 - \$35354 - indicates a broad range of monthly premiums, could be that they have a couple of high risk clients with majorly increased premiums
    - mean: 193.2
    - median: 83.0
    - std: 1601.2
    - *Stat Conclusion*: The mean premium is ~\$193, indicating that the average client pays a decently high monthly premium; the median being significantly lower indicates a major skew of higher premiums, caused probably by wealthier clients or high risk clients, and the std indicates a seemingly very wide spread of premiums at around ~1.6k. This indicates that they have a majority of low income clients with a few relatively more wealthy/higher-risk clients.

 - Total Claim : \$.38 - \$2893.24 - indicates a broad range of total claim
 amounts; seems they do not pay out much expensive claims .
  - mean: 404.98
    - median: 354.0
    - std: 293.02
    - *Stat Conclusion*: The mean premium is ~\$404, indicating that the average client gets a small claim from the insurance company; the median being slightly lower indicates a minor skew of higher premiums. The std at 293.02 is smaller than the mean, which may indicate a narrow range of claims made. This indicates that the car insurance complany does not pay out much to their clients in order to seek porfot.

 - customer lifetime value: All I can conclude is that their lifetime value percentages seem to be very high at 10 ** 5 factor.


## Challenge 2: analyzing the data

### **Numerical Data:**
**from reading the data types**:
It seems any categorical or string id columns are object type while the few numerical are float64.
2 Columns seem to have out of place dtypes:
- Number of Open Complaints: should be INT instead of object
- Customer Lifetime Value, which should be Float64 instead of object


In [15]:
clv = car_insurance_df['Customer Lifetime Value'].apply(lambda x: float(str(x).strip('%')))
car_insurance_df['Customer Lifetime Value'] = clv
print(car_insurance_df['Customer Lifetime Value'].dtype)
num_complaints = []
types = set()
for num in car_insurance_df['Number of Open Complaints']:
    if isinstance(num,str):
        num_complaints.append(int(num.split('/')[1]))
    else:
        num_complaints.append(-1)
    types.add(type(num))

car_insurance_df['Number of Open Complaints'] = num_complaints
car_insurance_df['Number of Open Complaints'].dtype

float64


dtype('int64')

In [9]:
car_insurance_df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       7
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [10]:
# Your code here
customer_locs = car_insurance_df['ST'].value_counts()
customer_locs['Arizona'] += customer_locs['AZ']
customer_locs['Washington'] += customer_locs['WA']
customer_locs['California'] += customer_locs['Cali']
customer_locs.drop(['AZ','Cali','WA'],inplace=True)
customer_locs.sort_values(inplace=True)
customer_locs.iloc[0:5]

ST
Nevada         98
Washington    111
Arizona       211
Oregon        320
California    331
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [11]:
# Your code here
policy_types = car_insurance_df['Policy Type'].value_counts()
print(policy_types, f"Highest sold is {policy_types.idxmax()}")

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64 Highest sold is Personal Auto


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [25]:
# Your code here
personal_df = car_insurance_df.loc[car_insurance_df['Policy Type'] == 'Personal Auto']
corporate_df = car_insurance_df.loc[car_insurance_df['Policy Type'] == 'Corporate Auto']

In [13]:
personal_mean = personal_df['Income'].mean()
corporate_mean = corporate_df['Income'].mean()
print(f"personal auto mean income : ${personal_mean:.2f}")
print(f"corporate auto mean income : ${corporate_mean:.2f}")

personal auto mean income : $38180.70
corporate auto mean income : $41390.31


### Bonus: Exercise 4


In [16]:
car_insurance_df.describe(include='number')['Total Claim Amount']

count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: Total Claim Amount, dtype: float64

Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data.

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [21]:
# Your code here
total_claim = car_insurance_df['Total Claim Amount']
top_25_perc = total_claim.quantile(.75)

In [22]:
high_policy_df = car_insurance_df[total_claim > top_25_perc]

In [24]:
high_policy_df.describe(include='number')['Total Claim Amount']

count     264.000000
mean      782.228263
std       292.751640
min       537.600000
25%       606.521741
50%       679.597985
75%       851.400000
max      2893.239678
Name: Total Claim Amount, dtype: float64